In [1]:
import pandas as pd
from datetime import datetime as dt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.metrics import accuracy_score, mean_absolute_error ,mean_squared_error, confusion_matrix, median_absolute_error,classification_report, f1_score,recall_score,precision_score
from sklearn.tree import DecisionTreeClassifier
%matplotlib inline



# 1- Loading & Splitting Data

In [2]:
# Load the training and testing data
train_data = pd.read_csv('fraudTrain.csv')
test_data = pd.read_csv('fraudTest.csv')

# Separate the features and labels
X_train = train_data.drop('is_fraud', axis=1)
y_train = train_data['is_fraud']
X_test = test_data.drop('is_fraud', axis=1)
y_test = test_data['is_fraud']

# 2- Insights about the data

In [3]:
train_data.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   

      last gender                        street  ...      lat      long  \
0    Banks      F                561 Perry Cove  ...  36.0788  -81.1781   
1     Gill      F  43039 Riley Greens Suite 393  ...  48.8878 -118.2105   
2  Sanchez      M      594 White Dale Suite 530  ...  42.1808 -112.2620   
3    White      M   9443 Cynthia Court Apt. 038  ...  46.2306 -112.1138   
4   Garcia      M              408 Bradley Rest  ...  38.4207  -79.4629   

   city_pop                                job         dob  \
0      3495          Psychologist, counselling  1988-03-09   
1       149  Special educational needs teacher  1978-06-21   
2      4154        Nature conservation officer  1962-01-19   
3      1939                    Patent attorney  1967-01-12   
4        99     Dance movement psychotherapist  1986-03-28   

                          trans_num   unix_time  merch_lat  merch_long  \
0  0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704 -112.154481   
3  6b849c168bdad6f867558c3793159a81  1325376076  47.034331 -112.561071   
4  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999  -78.632459   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [4]:
test_data.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  1968-03-19   
1       302  Sales professional, IT  1990-01-17   
2     34496       Librarian, public  1970-10-21   
3     54767            Set designer  1987-07-25   
4      1126      Furniture designer  1955-07-06   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [5]:
train_data.describe

<bound method NDFrame.describe of          Unnamed: 0 trans_date_trans_time               cc_num  \
0                 0   2019-01-01 00:00:18     2703186189652095   
1                 1   2019-01-01 00:00:44         630423337322   
2                 2   2019-01-01 00:00:51       38859492057661   
3                 3   2019-01-01 00:01:16     3534093764340240   
4                 4   2019-01-01 00:03:06      375534208663984   
...             ...                   ...                  ...   
1296670     1296670   2020-06-21 12:12:08       30263540414123   
1296671     1296671   2020-06-21 12:12:19     6011149206456997   
1296672     1296672   2020-06-21 12:12:32     3514865930894695   
1296673     1296673   2020-06-21 12:13:36     2720012583106919   
1296674     1296674   2020-06-21 12:13:37  4292902571056973207   

                                    merchant       category     amt  \
0                 fraud_Rippin, Kub and Mann       misc_net    4.97   
1            fraud_Heller, Gutm

In [6]:
test_data.describe

<bound method NDFrame.describe of         Unnamed: 0 trans_date_trans_time            cc_num  \
0                0   2020-06-21 12:14:25  2291163933867244   
1                1   2020-06-21 12:14:33  3573030041201292   
2                2   2020-06-21 12:14:53  3598215285024754   
3                3   2020-06-21 12:15:15  3591919803438423   
4                4   2020-06-21 12:15:17  3526826139003047   
...            ...                   ...               ...   
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                                    merchant        category     amt    first  \
0                      fraud_Kirlin and Sons   personal_care    2.86     Jeff   
1                       fraud_Sporer-Keebler   personal_car

In [ ]:
train_data.isna()
test_data.isna()

In [7]:
train_data.dtypes

Unnamed: 0                 int64
trans_date_trans_time     object
cc_num                     int64
merchant                  object
category                  object
amt                      float64
first                     object
last                      object
gender                    object
street                    object
city                      object
state                     object
zip                        int64
lat                      float64
long                     float64
city_pop                   int64
job                       object
dob                       object
trans_num                 object
unix_time                  int64
merch_lat                float64
merch_long               float64
is_fraud                   int64
dtype: object

In [8]:
test_data.dtypes

Unnamed: 0                 int64
trans_date_trans_time     object
cc_num                     int64
merchant                  object
category                  object
amt                      float64
first                     object
last                      object
gender                    object
street                    object
city                      object
state                     object
zip                        int64
lat                      float64
long                     float64
city_pop                   int64
job                       object
dob                       object
trans_num                 object
unix_time                  int64
merch_lat                float64
merch_long               float64
is_fraud                   int64
dtype: object

# 3- Preprocessing Data

In [9]:
#drop unnecessary columns for both train and test 
train_data.drop("Unnamed: 0",axis=1,inplace=True) 
test_data.drop("Unnamed: 0",axis=1,inplace=True) 
train_data = train_data.drop(['cc_num','first','last','trans_num'],axis=1)
test_data = test_data.drop(['cc_num','first','last','trans_num'],axis=1)

In [10]:
#adding trans_date column to extract only the date from trans_date_trans_time column in test data
train_data["trans_date_trans_time"] = pd.to_datetime(train_data["trans_date_trans_time"])
train_data["trans_date"] = train_data["trans_date_trans_time"].dt.date
train_data["trans_date"]= pd.to_datetime(train_data["trans_date"])


In [11]:
#extracting month,year,day and adding a column for them from trans_date from train dataset
train_data['year'] = train_data['trans_date'].dt.year
train_data['month'] =train_data['trans_date'].dt.month
train_data['day'] = train_data['trans_date'].dt.day

In [12]:
train_data

trans_date_trans_time                             merchant  \
0         2019-01-01 00:00:18           fraud_Rippin, Kub and Mann   
1         2019-01-01 00:00:44      fraud_Heller, Gutmann and Zieme   
2         2019-01-01 00:00:51                 fraud_Lind-Buckridge   
3         2019-01-01 00:01:16   fraud_Kutch, Hermiston and Farrell   
4         2019-01-01 00:03:06                  fraud_Keeling-Crist   
...                       ...                                  ...   
1296670   2020-06-21 12:12:08                    fraud_Reichel Inc   
1296671   2020-06-21 12:12:19             fraud_Abernathy and Sons   
1296672   2020-06-21 12:12:32                 fraud_Stiedemann Ltd   
1296673   2020-06-21 12:13:36  fraud_Reinger, Weissnat and Strosin   
1296674   2020-06-21 12:13:37  fraud_Langosh, Wintheiser and Hyatt   

              category     amt gender                         street  \
0             misc_net    4.97      F                 561 Perry Cove   
1          grocery_pos  107.23      F   43039 Riley Greens Suite 393   
2        entertainment  220.11      M       594 White Dale Suite 530   
3        gas_transport   45.00      M    9443 Cynthia Court Apt. 038   
4             misc_pos   41.96      M               408 Bradley Rest   
...                ...     ...    ...                            ...   
1296670  entertainment   15.56      M       162 Jessica Row Apt. 072   
1296671    food_dining   51.70      M  8617 Holmes Terrace Suite 651   
1296672    food_dining  105.93      M     1632 Cohen Drive Suite 639   
1296673    food_dining   74.90      M           42933 Ryan Underpass   
1296674    food_dining    4.30      M           135 Joseph Mountains   

                             city state    zip      lat  ...  \
0                  Moravian Falls    NC  28654  36.0788  ...   
1                          Orient    WA  99160  48.8878  ...   
2                      Malad City    ID  83252  42.1808  ...   
3                         Boulder    MT  59632  46.2306  ...   
4                        Doe Hill    VA  24433  38.4207  ...   
...                           ...   ...    ...      ...  ...   
1296670                     Hatch    UT  84735  37.7175  ...   
1296671                 Tuscarora    MD  21790  39.2667  ...   
1296672  High Rolls Mountain Park    NM  88325  32.9396  ...   
1296673                 Manderson    SD  57756  43.3526  ...   
1296674                      Sula    MT  59871  45.8433  ...   

                                       job         dob   unix_time  merch_lat  \
0                Psychologist, counselling  1988-03-09  1325376018  36.011293   
1        Special educational needs teacher  1978-06-21  1325376044  49.159047   
2              Nature conservation officer  1962-01-19  1325376051  43.150704   
3                          Patent attorney  1967-01-12  1325376076  47.034331   
4           Dance movement psychotherapist  1986-03-28  1325376186  38.674999   
...                                    ...         ...         ...        ...   
1296670                       Geoscientist  1961-11-24  1371816728  36.841266   
1296671   Production assistant, television  1979-12-11  1371816739  38.906881   
1296672                    Naval architect  1967-08-30  1371816752  33.619513   
1296673              Volunteer coordinator  1980-08-18  1371816816  42.788940   
1296674           Therapist, horticultural  1995-08-16  1371816817  46.565983   

         merch_long  is_fraud  trans_date  year month  day  
0        -82.048315         0  2019-01-01  2019     1    1  
1       -118.186462         0  2019-01-01  2019     1    1  
2       -112.154481         0  2019-01-01  2019     1    1  
3       -112.561071         0  2019-01-01  2019     1    1  
4        -78.632459         0  2019-01-01  2019     1    1  
...             ...       ...         ...   ...   ...  ...  
1296670 -111.690765         0  2020-06-21  2020     6   21  
1296671  -78.246528         0  2020-06-21  2020     6   21  
1296672 -10

In [13]:
#adding trans_date column to extract only the date from trans_date_trans_time column in test data
test_data["trans_date_trans_time"] = pd.to_datetime(test_data["trans_date_trans_time"])
test_data["trans_date"] = test_data["trans_date_trans_time"].dt.date
test_data["trans_date"]= pd.to_datetime(test_data["trans_date"]) 

In [14]:
#extracting month,year,day and adding a column for them from trans_date from test dataset
test_data['year'] = test_data['trans_date'].dt.year
test_data['month'] = test_data['trans_date'].dt.month
test_data['day'] = test_data['trans_date'].dt.day

In [15]:
test_data

trans_date_trans_time                              merchant  \
0        2020-06-21 12:14:25                 fraud_Kirlin and Sons   
1        2020-06-21 12:14:33                  fraud_Sporer-Keebler   
2        2020-06-21 12:14:53  fraud_Swaniawski, Nitzsche and Welch   
3        2020-06-21 12:15:15                     fraud_Haley Group   
4        2020-06-21 12:15:17                 fraud_Johnston-Casper   
...                      ...                                   ...   
555714   2020-12-31 23:59:07                 fraud_Reilly and Sons   
555715   2020-12-31 23:59:09                  fraud_Hoppe-Parisian   
555716   2020-12-31 23:59:15                       fraud_Rau-Robel   
555717   2020-12-31 23:59:24                 fraud_Breitenberg LLC   
555718   2020-12-31 23:59:34                     fraud_Dare-Marvin   

              category     amt gender                       street  \
0        personal_care    2.86      M            351 Darlene Green   
1        personal_care   29.84      F             3638 Marsh Union   
2       health_fitness   41.28      F         9333 Valentine Point   
3             misc_pos   60.05      M  32941 Krystal Mill Apt. 552   
4               travel    3.19      M     5783 Evan Roads Apt. 465   
...                ...     ...    ...                          ...   
555714  health_fitness   43.77      M          558 Michael Estates   
555715       kids_pets  111.84      M          572 Davis Mountains   
555716       kids_pets   86.88      F   144 Evans Islands Apt. 683   
555717          travel    7.99      M   7020 Doyle Stream Apt. 951   
555718   entertainment   38.13      M     830 Myers Plaza Apt. 384   

                city state    zip      lat  ...                     job  \
0           Columbia    SC  29209  33.9659  ...     Mechanical engineer   
1            Altonah    UT  84002  40.3207  ...  Sales professional, IT   
2           Bellmore    NY  11710  40.6729  ...       Librarian, public   
3         Titusville    FL  32780  28.5697  ...            Set designer   
4           Falmouth    MI  49632  44.2529  ...      Furniture designer   
...              ...   ...    ...      ...  ...                     ...   
555714         Luray    MO  63453  40.4931  ...            Town planner   
555715  Lake Jackson    TX  77566  29.0393  ...          Futures trader   
555716       Burbank    WA  99323  46.1966  ...                Musician   
555717          Mesa    ID  83643  44.6255  ...            Cartographer   
555718        Edmond    OK  73034  35.6665  ...             Media buyer   

               dob   unix_time  merch_lat  merch_long  is_fraud  trans_date  \
0       1968-03-19  1371816865  33.986391  -81.200714         0  2020-06-21   
1       1990-01-17  1371816873  39.450498 -109.960431         0  2020-06-21   
2       1970-10-21  1371816893  40.495810  -74.196111         0  2020-06-21   
3       1987-07-25  1371816915  28.812398  -80.883061         0  2020-06-21   
4       1955-07-06  1371816917  44.959148  -85.884734         0  2020-06-21   
...            ...         ...        ...         ...       ...         ...   
555714  1966-02-13  1388534347  39.946837  -91.333331         0  2020-12-31   
555715  1999-12-27  1388534349  29.661049  -96.186633         0  2020-12-31   
555716  1981-11-29  1388534355  46.658340 -119.715054         0  2020-12-31   
555717  1965-12-15  1388534364  44.470525 -117.080888         0  2020-12-31   
555718  1993-05-10  1388534374  36.210097  -97.036372         0  2020-12-31   

        year month  day  
0       2020     6   21  
1       2020     6   21  
2       2020     6   21  
3       2020     6   21  
4       2020     6   21  
...      ...   ...  ...  
555714  2020    12   31  
555715  2020    12   31  
555716  2020    12   31  
555717  2020    12   31  
555718  2020    12   31  

[555719 rows x 22 columns]

In [16]:
#fraud count
train_data["is_fraud"].value_counts()

0    1289169
1       7506
Name: is_fraud, dtype: int64

So there is 7506 fraud transactions in training set

In [17]:
#fraud count in test data
test_data["is_fraud"].value_counts()

0    553574
1      2145
Name: is_fraud, dtype: int64

So there is 2145 fraud transactions in test set

In [18]:
train_data.groupby("is_fraud")['amt'].mean()

is_fraud
0     67.667110
1    531.320092
Name: amt, dtype: float64

In [19]:
pd.crosstab(train_data["category"],train_data["is_fraud"])

is_fraud             0     1
category                    
entertainment    93781   233
food_dining      91310   151
gas_transport   131041   618
grocery_net      45318   134
grocery_pos     121895  1743
health_fitness   85746   133
home            122917   198
kids_pets       112796   239
misc_net         62372   915
misc_pos         79405   250
personal_care    90538   220
shopping_net     95830  1713
shopping_pos    115829   843
travel           40391   116

In [20]:
#encode categorical data
train_data = pd.get_dummies(train_data,columns=['category'],drop_first=True)
test_data = pd.get_dummies(test_data,columns=['category'],drop_first=True)

In [21]:
#remove spaces 
train_data.columns = train_data.columns.str.replace(' ', '')
test_data.columns = test_data.columns.str.replace(' ', '')

In [22]:
#we make data set from only number data types to put in model
train = train_data.select_dtypes(include='number')
test = test_data.select_dtypes(include='number')

In [23]:
#concatanation operation for x and y
total = pd.concat([train, test])

In [24]:
total

amt    zip      lat      long  city_pop   unix_time  merch_lat  \
0         4.97  28654  36.0788  -81.1781      3495  1325376018  36.011293   
1       107.23  99160  48.8878 -118.2105       149  1325376044  49.159047   
2       220.11  83252  42.1808 -112.2620      4154  1325376051  43.150704   
3        45.00  59632  46.2306 -112.1138      1939  1325376076  47.034331   
4        41.96  24433  38.4207  -79.4629        99  1325376186  38.674999   
...        ...    ...      ...       ...       ...         ...        ...   
555714   43.77  63453  40.4931  -91.8912       519  1388534347  39.946837   
555715  111.84  77566  29.0393  -95.4401     28739  1388534349  29.661049   
555716   86.88  99323  46.1966 -118.9017      3684  1388534355  46.658340   
555717    7.99  83643  44.6255 -116.4493       129  1388534364  44.470525   
555718   38.13  73034  35.6665  -97.4798    116001  1388534374  36.210097   

        merch_long  is_fraud  year  ...  category_grocery_pos  \
0       -82.048315         0  2019  ...                     0   
1      -118.186462         0  2019  ...                     1   
2      -112.154481         0  2019  ...                     0   
3      -112.561071         0  2019  ...                     0   
4       -78.632459         0  2019  ...                     0   
...            ...       ...   ...  ...                   ...   
555714  -91.333331         0  2020  ...                     0   
555715  -96.186633         0  2020  ...                     0   
555716 -119.715054         0  2020  ...                     0   
555717 -117.080888         0  2020  ...                     0   
555718  -97.036372         0  2020  ...                     0   

        category_health_fitness  category_home  category_kids_pets  \
0                             0              0                   0   
1                             0              0                   0   
2                             0              0                   0   
3                             0              0                   0   
4                             0              0                   0   
...                         ...            ...                 ...   
555714                        1              0                   0   
555715                        0              0                   1   
555716                        0              0                   1   
555717                        0              0                   0   
555718                        0              0                   0   

        category_misc_net  category_misc_pos  category_personal_care  \
0                       1                  0                       0   
1                       0                  0                       0   
2                       0                  0                       0   
3                       0                  0                       0   
4                       0                  1                       0   
...                   ...                ...                     ...   
555714                  0                  0                       0   
555715                  0                  0                       0   
555716                  0                  0                       0   
555717                  0                  0                       0   
555718                  0                  0                       0   

        category_shopping_net  category_shopping_pos  category_travel  
0                           0                      0                0  
1                           0                      0                0  
2                           0                      0                0  
3                           0                      0                0  
4                           0                      0                0  
...                       ...                    ...              ...  
555714                      0                      0                0  
555715                 

In [25]:
#X for features and y for output label
X = total.drop("is_fraud",axis=1) 
y = total["is_fraud"]

In [26]:
#sum of fraud in both test and train data set
print("Sum of fraud transactions in all dataset =",sum(y))

Sum of fraud transactions in all dataset = 9651


In [27]:
X

amt    zip      lat      long  city_pop   unix_time  merch_lat  \
0         4.97  28654  36.0788  -81.1781      3495  1325376018  36.011293   
1       107.23  99160  48.8878 -118.2105       149  1325376044  49.159047   
2       220.11  83252  42.1808 -112.2620      4154  1325376051  43.150704   
3        45.00  59632  46.2306 -112.1138      1939  1325376076  47.034331   
4        41.96  24433  38.4207  -79.4629        99  1325376186  38.674999   
...        ...    ...      ...       ...       ...         ...        ...   
555714   43.77  63453  40.4931  -91.8912       519  1388534347  39.946837   
555715  111.84  77566  29.0393  -95.4401     28739  1388534349  29.661049   
555716   86.88  99323  46.1966 -118.9017      3684  1388534355  46.658340   
555717    7.99  83643  44.6255 -116.4493       129  1388534364  44.470525   
555718   38.13  73034  35.6665  -97.4798    116001  1388534374  36.210097   

        merch_long  year  month  ...  category_grocery_pos  \
0       -82.048315  2019      1  ...                     0   
1      -118.186462  2019      1  ...                     1   
2      -112.154481  2019      1  ...                     0   
3      -112.561071  2019      1  ...                     0   
4       -78.632459  2019      1  ...                     0   
...            ...   ...    ...  ...                   ...   
555714  -91.333331  2020     12  ...                     0   
555715  -96.186633  2020     12  ...                     0   
555716 -119.715054  2020     12  ...                     0   
555717 -117.080888  2020     12  ...                     0   
555718  -97.036372  2020     12  ...                     0   

        category_health_fitness  category_home  category_kids_pets  \
0                             0              0                   0   
1                             0              0                   0   
2                             0              0                   0   
3                             0              0                   0   
4                             0              0                   0   
...                         ...            ...                 ...   
555714                        1              0                   0   
555715                        0              0                   1   
555716                        0              0                   1   
555717                        0              0                   0   
555718                        0              0                   0   

        category_misc_net  category_misc_pos  category_personal_care  \
0                       1                  0                       0   
1                       0                  0                       0   
2                       0                  0                       0   
3                       0                  0                       0   
4                       0                  1                       0   
...                   ...                ...                     ...   
555714                  0                  0                       0   
555715                  0                  0                       0   
555716                  0                  0                       0   
555717                  0                  0                       0   
555718                  0                  0                       0   

        category_shopping_net  category_shopping_pos  category_travel  
0                           0                      0                0  
1                           0                      0                0  
2                           0                      0                0  
3                           0                      0                0  
4                           0                      0                0  
...                       ...                    ...              ...  
555714                      0                      0                0  
555715                      0                      0       

In [28]:
#check data type for X 
X.dtypes

amt                        float64
zip                          int64
lat                        float64
long                       float64
city_pop                     int64
unix_time                    int64
merch_lat                  float64
merch_long                 float64
year                         int64
month                        int64
day                          int64
category_food_dining         uint8
category_gas_transport       uint8
category_grocery_net         uint8
category_grocery_pos         uint8
category_health_fitness      uint8
category_home                uint8
category_kids_pets           uint8
category_misc_net            uint8
category_misc_pos            uint8
category_personal_care       uint8
category_shopping_net        uint8
category_shopping_pos        uint8
category_travel              uint8
dtype: object

In [29]:
#drop columns that are irrelevant
X = X.drop(['zip','lat','long','unix_time','merch_lat','merch_long'],axis=1)

# 4-Building the Classification Model

In [30]:
from sklearn.model_selection import train_test_split
X_train1, X_test1, y_train1, y_test1 = train_test_split(
 X, y, test_size=0.3, random_state=42)

In [31]:
#scale data into range
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X_train1 = scaler.fit_transform(X_train1)
X_test1 = scaler.transform(X_test1)

Decision Tree model

In [32]:
dtc = DecisionTreeClassifier()
dtc.fit(X_train1,y_train1)

DecisionTreeClassifier()

Model evaluation

In [33]:
print("Score the X-train with Y-train is : ", dtc.score(X_train1,y_train1))
print("Score the X-test  with Y-test  is : ", dtc.score(X_test1,y_test1))

y_pred=dtc.predict(X_test1)

Score the X-train with Y-train is :  1.0
Score the X-test  with Y-test  is :  0.9958756133945393


In [34]:
print("Accuracy score " , accuracy_score(y_test1,y_pred))
print("F1 score: ", round(f1_score(y_test1, y_pred, average='weighted')*100,2),"%")

Accuracy score  0.9958756133945393
F1 score:  99.59 %


In [35]:
matrix = classification_report(y_test1,y_pred,labels=[1,0])
print('Classification report : \n',matrix)

Classification report : 
               precision    recall  f1-score   support

           1       0.60      0.63      0.62      2938
           0       1.00      1.00      1.00    552781

    accuracy                           1.00    555719
   macro avg       0.80      0.82      0.81    555719
weighted avg       1.00      1.00      1.00    555719



In [36]:
print('Precision: %.3f' % precision_score(y_test1, y_pred))


Precision: 0.605


In [37]:
print("F1 score: ", round(f1_score(y_test1, y_pred, average='weighted')*100,2),"%")

F1 score:  99.59 %


Random Forest Model

In [38]:
rfc = RandomForestClassifier(n_estimators=100, random_state=42)

In [40]:
# Fit the model to the training data
rfc.fit(X_train1, y_train1)

RandomForestClassifier(random_state=42)

In [41]:
# Make predictions on the test data
y_pred1 = rfc.predict(X_test1)

Model evaluation

In [42]:
print("Score the X-train with Y-train is : ", rfc.score(X_train1,y_train1))
print("Score the X-test  with Y-test  is : ", rfc.score(X_test1,y_test1))

Score the X-train with Y-train is :  0.9999992287967301
Score the X-test  with Y-test  is :  0.9974969364013108


In [43]:
print('Classification report : \n',classification_report(y_test1,y_pred1,labels=[1,0]))

Classification report : 
               precision    recall  f1-score   support

           1       0.84      0.65      0.73      2938
           0       1.00      1.00      1.00    552781

    accuracy                           1.00    555719
   macro avg       0.92      0.83      0.87    555719
weighted avg       1.00      1.00      1.00    555719



In [44]:
print("Accuracy score " , accuracy_score(y_test1,y_pred1))
print("F1 score: ", round(f1_score(y_test1, y_pred1, average='weighted')*100,2),"%")

Accuracy score  0.9974969364013108
F1 score:  99.73 %


Logistic regression

In [45]:
from sklearn import linear_model
lgr=linear_model.LogisticRegression()

In [46]:
lgr.fit(X_train1, y_train1)

LogisticRegression()

In [47]:
y_pred2 = lgr.predict(X_test1)

In [48]:
print('Classification report : \n',classification_report(y_test1,y_pred2,labels=[1,0]))

Classification report : 
               precision    recall  f1-score   support

           1       0.00      0.00      0.00      2938
           0       0.99      1.00      1.00    552781

    accuracy                           0.99    555719
   macro avg       0.50      0.50      0.50    555719
weighted avg       0.99      0.99      0.99    555719



In [ ]:
print("Accuracy score " , accuracy_score(y_test1,y_pred2))
print("F1 score: ", round(f1_score(y_test1, y_pred2, average='weighted')*100,2),"%")

Wohoo The random forest model has the biggest accuracy!